File: SharedKit/Sources/SharedKit/Utilities/Cache.swift
import Foundation

/// Thread-safe generic cache with TTL support
public final class Cache<Key: Hashable, Value> {
    
    private struct CacheEntry {
        let value: Value
        let expirationDate: Date?
        
        var isExpired: Bool {
            guard let expiration = expirationDate else { return false }
            return Date() > expiration
        }
    }
    
    private var storage: [Key: CacheEntry] = [:]
    private let lock = NSLock()
    private let defaultTTL: TimeInterval?
    private let maxEntries: Int?
    
    /// Creates a new cache
    /// - Parameters:
    ///   - defaultTTL: Default time-to-live for entries (nil = no expiration)
    ///   - maxEntries: Maximum number of entries (nil = unlimited)
    public init(defaultTTL: TimeInterval? = nil, maxEntries: Int? = nil) {
        self.defaultTTL = defaultTTL
        self.maxEntries = maxEntries
    }
    
    /// Gets a value from the cache
    public func get(_ key: Key) -> Value? {
        lock.lock()
        defer { lock.unlock() }
        
        guard let entry = storage[key] else { return nil }
        
        if entry.isExpired {
            storage.removeValue(forKey: key)
            return nil
        }
        
        return entry.value
    }
    
    /// Sets a value in the cache
    /// - Parameters:
    ///   - value: The value to cache
    ///   - key: The cache key
    ///   - ttl: Time-to-live for this entry (overrides default)
    public func set(_ value: Value, forKey key: Key, ttl: TimeInterval? = nil) {
        lock.lock()
        defer { lock.unlock() }
        
        let effectiveTTL = ttl ?? defaultTTL
        let expiration = effectiveTTL.map { Date().addingTimeInterval($0) }
        
        storage[key] = CacheEntry(value: value, expirationDate: expiration)
        
        // Evict oldest entries if over limit
        if let max = maxEntries, storage.count > max {
            evictOldestEntries(keepCount: max)
        }
    }
    
    /// Gets a value or computes and caches it if missing
    public func getOrCompute(_ key: Key, ttl: TimeInterval? = nil, compute: () -> Value) -> Value {
        if let cached = get(key) {
            return cached
        }
        
        let value = compute()
        set(value, forKey: key, ttl: ttl)
        return value
    }
    
    /// Async version of getOrCompute
    public func getOrCompute(_ key: Key, ttl: TimeInterval? = nil, compute: () async -> Value) async -> Value {
        if let cached = get(key) {
            return cached
        }
        
        let value = await compute()
        set(value, forKey: key, ttl: ttl)
        return value
    }
    
    /// Removes a value from the cache
    public func remove(_ key: Key) {
        lock.lock()
        defer { lock.unlock() }
        storage.removeValue(forKey: key)
    }
    
    /// Clears all entries from the cache
    public func clear() {
        lock.lock()
        defer { lock.unlock() }
        storage.removeAll()
    }
    
    /// Removes all expired entries
    public func purgeExpired() {
        lock.lock()
        defer { lock.unlock() }
        
        let keysToRemove = storage.filter { $0.value.isExpired }.map { $0.key }
        for key in keysToRemove {
            storage.removeValue(forKey: key)
        }
    }
    
    /// Current number of entries
    public var count: Int {
        lock.lock()
        defer { lock.unlock() }
        return storage.count
    }
    
    /// Check if a key exists (and is not expired)
    public func contains(_ key: Key) -> Bool {
        return get(key) != nil
    }
    
    private func evictOldestEntries(keepCount: Int) {
        // Simple eviction: remove entries until under limit
        // In a production app, you might want LRU eviction
        while storage.count > keepCount {
            if let firstKey = storage.keys.first {
                storage.removeValue(forKey: firstKey)
            }
        }
    }
}

// MARK: - Convenience Extensions

public extension Cache where Key == String {
    /// Gets or computes with automatic key generation from parameters
    func cached<P: CustomStringConvertible>(
        for params: P,
        ttl: TimeInterval? = nil,
        compute: () -> Value
    ) -> Value {
        let key = String(describing: params)
        return getOrCompute(key, ttl: ttl, compute: compute)
    }
}

// MARK: - Memoization Helper

/// Memoizes a function with a single hashable argument
public func memoize<Input: Hashable, Output>(
    ttl: TimeInterval? = nil,
    _ function: @escaping (Input) -> Output
) -> (Input) -> Output {
    let cache = Cache<Input, Output>(defaultTTL: ttl)
    return { input in
        cache.getOrCompute(input) { function(input) }
    }
}

/// Memoizes an async function with a single hashable argument
public func memoizeAsync<Input: Hashable, Output>(
    ttl: TimeInterval? = nil,
    _ function: @escaping (Input) async -> Output
) -> (Input) async -> Output {
    let cache = Cache<Input, Output>(defaultTTL: ttl)
    return { input in
        await cache.getOrCompute(input) { await function(input) }
    }
}

// MARK: - Computed Property Cache

/// Cache for expensive computed properties
public final class ComputedCache<Value> {
    private var cachedValue: Value?
    private var lastComputedAt: Date?
    private let ttl: TimeInterval?
    private let lock = NSLock()
    
    public init(ttl: TimeInterval? = nil) {
        self.ttl = ttl
    }
    
    public func get(compute: () -> Value) -> Value {
        lock.lock()
        defer { lock.unlock() }
        
        // Check if cached value is still valid
        if let value = cachedValue {
            if let ttl = ttl, let lastComputed = lastComputedAt {
                if Date().timeIntervalSince(lastComputed) < ttl {
                    return value
                }
            } else if ttl == nil {
                return value
            }
        }
        
        // Compute and cache
        let value = compute()
        cachedValue = value
        lastComputedAt = Date()
        return value
    }
    
    public func invalidate() {
        lock.lock()
        defer { lock.unlock() }
        cachedValue = nil
        lastComputedAt = nil
    }
}

File: NotchToDo-Android/app/src/main/java/com/notchtodo/util/Cache.kt
package com.notchtodo.util

import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit

/**
 * Thread-safe generic cache with TTL support.
 */
class Cache<K : Any, V : Any>(
    private val defaultTTLMillis: Long? = null,
    private val maxEntries: Int? = null
) {
    private data class CacheEntry<V>(
        val value: V,
        val expirationTime: Long?
    ) {
        val isExpired: Boolean
            get() = expirationTime != null && System.currentTimeMillis() > expirationTime
    }
    
    private val storage = ConcurrentHashMap<K, CacheEntry<V>>()
    
    /**
     * Gets a value from the cache.
     */
    fun get(key: K): V? {
        val entry = storage[key] ?: return null
        
        if (entry.isExpired) {
            storage.remove(key)
            return null
        }
        
        return entry.value
    }
    
    /**
     * Sets a value in the cache.
     */
    fun set(key: K, value: V, ttlMillis: Long? = null) {
        val effectiveTTL = ttlMillis ?: defaultTTLMillis
        val expiration = effectiveTTL?.let { System.currentTimeMillis() + it }
        
        storage[key] = CacheEntry(value, expiration)
        
        // Evict if over limit
        maxEntries?.let { max ->
            while (storage.size > max) {
                storage.keys.firstOrNull()?.let { storage.remove(it) }
            }
        }
    }
    
    /**
     * Gets a value or computes and caches it if missing.
     */
    inline fun getOrCompute(key: K, ttlMillis: Long? = null, compute: () -> V): V {
        get(key)?.let { return it }
        
        val value = compute()
        set(key, value, ttlMillis)
        return value
    }
    
    /**
     * Async version of getOrCompute.
     */
    suspend inline fun getOrComputeAsync(
        key: K,
        ttlMillis: Long? = null,
        crossinline compute: suspend () -> V
    ): V {
        get(key)?.let { return it }
        
        val value = compute()
        set(key, value, ttlMillis)
        return value
    }
    
    /**
     * Removes a value from the cache.
     */
    fun remove(key: K) {
        storage.remove(key)
    }
    
    /**
     * Clears all entries from the cache.
     */
    fun clear() {
        storage.clear()
    }
    
    /**
     * Removes all expired entries.
     */
    fun purgeExpired() {
        val keysToRemove = storage.filter { it.value.isExpired }.keys
        keysToRemove.forEach { storage.remove(it) }
    }
    
    /**
     * Current number of entries.
     */
    val size: Int get() = storage.size
    
    /**
     * Check if a key exists (and is not expired).
     */
    fun contains(key: K): Boolean = get(key) != null
    
    companion object {
        /**
         * Creates a cache with TTL in seconds.
         */
        fun <K : Any, V : Any> withTTLSeconds(
            seconds: Long,
            maxEntries: Int? = null
        ): Cache<K, V> {
            return Cache(TimeUnit.SECONDS.toMillis(seconds), maxEntries)
        }
        
        /**
         * Creates a cache with TTL in minutes.
         */
        fun <K : Any, V : Any> withTTLMinutes(
            minutes: Long,
            maxEntries: Int? = null
        ): Cache<K, V> {
            return Cache(TimeUnit.MINUTES.toMillis(minutes), maxEntries)
        }
    }
}

/**
 * Memoizes a function with a single argument.
 */
fun <I : Any, O : Any> memoize(
    ttlMillis: Long? = null,
    function: (I) -> O
): (I) -> O {
    val cache = Cache<I, O>(ttlMillis)
    return { input -> cache.getOrCompute(input) { function(input) } }
}

/**
 * Cache for expensive computed properties.
 */
class ComputedCache<V : Any>(
    private val ttlMillis: Long? = null
) {
    @Volatile
    private var cachedValue: V? = null
    
    @Volatile
    private var lastComputedAt: Long? = null
    
    private val lock = Any()
    
    fun get(compute: () -> V): V {
        synchronized(lock) {
            // Check if cached value is still valid
            cachedValue?.let { value ->
                val ttl = ttlMillis
                val lastComputed = lastComputedAt
                
                if (ttl == null) {
                    return value
                }
                
                if (lastComputed != null && System.currentTimeMillis() - lastComputed < ttl) {
                    return value
                }
            }
            
            // Compute and cache
            val value = compute()
            cachedValue = value
            lastComputedAt = System.currentTimeMillis()
            return value
        }
    }
    
    suspend fun getAsync(compute: suspend () -> V): V {
        // Check cache first without lock
        cachedValue?.let { value ->
            val ttl = ttlMillis
            val lastComputed = lastComputedAt
            
            if (ttl == null) {
                return value
            }
            
            if (lastComputed != null && System.currentTimeMillis() - lastComputed < ttl) {
                return value
            }
        }
        
        // Compute (potentially expensive)
        val value = compute()
        
        // Update cache with lock
        synchronized(lock) {
            cachedValue = value
            lastComputedAt = System.currentTimeMillis()
        }
        
        return value
    }
    
    fun invalidate() {
        synchronized(lock) {
            cachedValue = null
            lastComputedAt = null
        }
    }
}
